src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -676,7 +676,11 @@ protected synchronized void initializeStorage() throws IOException {
                 h.setForDeserialize(storage.getNode(entry.getValue()));
                 heads.put(h.getId(), h);
             } else {
-                throw new IOException("Tried to load head FlowNodes for execution "+this.owner+" but FlowNode was not found in storage for head id:FlowNodeId "+entry.getKey()+":"+entry.getValue());
+                String suggestDurability  = "";
+                if (getDurabilityHint() != FlowDurabilityHint.MAX_SURVIVABILITY) {
+                    suggestDurability =  "Try setting pipeline durability to MAX_SURVIVABILITY. ";
+                }
+                throw new IOException("Cannot resume build because Jenkins did not shut down cleanly. " + suggestDurability + (LOGGER.isLoggable(Level.FINE) ? "(Missing FlowNode " + entry.getValue() + " for FlowHead " + entry.getKey() + ")" : ""));
```

Comment:
This is not correct.  Jenkins may have shut down correctly, we can not know.

All we know is a file we expect to be there is not.  This could be disk corruption outside Jenkins or a restore from a non atomic backup....

We should say we can not find a file (and which file, log it add a chained FileNotFoundExcption?).  If you are not using a durable mode we should say to use it, otherwise this is expected for a non clean shutdown.

Reply:
In my opinion, the specific XML file that is missing is completely irrelevant to users (and probably even to plugin developers). I would guess that 95% or more of the time users only see this exception when Jenkins crashed while using the `PERFORMANCE_OPTIMIZED` durability setting, and in that scenario the new message is strictly an improvement over the old one since the behavior is typical and expected.

If you really want to improve the logic to diagnose the scenario more precisely, then I would recommend looking into https://github.com/jenkinsci/workflow-cps-plugin/pull/363 or https://github.com/jenkinsci/workflow-cps-plugin/compare/master...dwnusbaum:JENKINS-55287-2 like I mentioned in https://github.com/jenkinsci/workflow-cps-plugin/pull/589#pullrequestreview-1122294880, but it gets complex fast because this code is undertested and there are many subtle issues with the existing logic and its interaction with `WorkflowRun.onLoad`, and I am not confident that either of those changes would fix the existing bugs without introducing new issues, which is why I abandoned them.

If you want, you could keep the existing logic more or less but make the message a lot more detailed and hedge all recommendations, but in practice IDK if it really makes a difference: 

```
if (durabilitySetting != MAX_SURVIVABILITY) {
    throw new AbortException("Cannot resume build because flow node X could not be loaded. " +
            "This is expected to happen when using the " + durabilitySetting + " durability setting and Jenkins is not shut down cleanly. " +
            "Consider investigating to understand if Jenkins was not shut down cleanly or switching to the MAX_SURVIVABILITY durability setting which should prevent this issue in most cases." );
} else {
    throw new AbortException("Cannot resume build because flow node X could not be loaded.");
}
```